lora/models.py

```diff
@@ -32,14 +90,58 @@ def from_linear(linear: nn.Linear, rank: int = 8):
         lora_lin.linear = linear
         return lora_lin
 
+    def to_linear(self):
+        linear = self.linear
+        bias = "bias" in linear
+        weight = linear.weight
+        is_quantized = isinstance(linear, nn.QuantizedLinear)
+
+        # Use the same type as the linear weight if not quantized
+        dtype = weight.dtype
+
+        if is_quantized:
+            dtype = mx.float16
+            weight = mx.dequantize(
+                weight,
+                linear.scales,
+                linear.biases,
+                linear.group_size,
+                linear.bits,
+            )
+        output_dims, input_dims = weight.shape
+        fused_linear = nn.Linear(input_dims, output_dims, bias=bias)
+
+        lora_b = (self.scale * self.lora_b.T).astype(dtype)
+        lora_a = self.lora_a.T.astype(dtype)
+        fused_linear.weight = weight + lora_b @ lora_a
+        if bias:
+            fused_linear.bias = linear.bias
+
+        if is_quantized:
+            fused_linear = nn.QuantizedLinear.from_linear(
+                fused_linear,
+                linear.group_size,
+                linear.bits,
+            )
+
+        return fused_linear
+
     def __init__(
-        self, input_dims: int, output_dims: int, lora_rank: int = 8, bias: bool = False
+        self,
+        input_dims: int,
+        output_dims: int,
+        lora_rank: int = 8,
+        bias: bool = False,
+        scale: float = 20.0,
```

Comment:
More of a curiosity, is the 20 a better default than 2?

Reply:
It trains much faster with QLoRA and fusing didn’t work at all with a smaller scale. The low rank update was getting completely wiped out in the dequantize - requantize part.

We should let it be configurable as an arg, but so far 20 was consistently better in the cases I tried so it’s a step in the right direction.

Reply:
I have noticed in the past that when I try to merge lora weight using transformers' peft library, I had some performance degradation. Now I am wondering if it was caused by this issue. However, by quickly checking the source code, it seems like they are using a quick small scaling (16/64). I'm not sure how it works. 
https://github.com/artidoro/qlora/blob/main/qlora.py#L177-L185
https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/layer.py#L90-L93

Reply:
That's interesting 🤔  .. maybe they initialize the adapters differently?

Reply:
I will read their code a bit more today to see if I can find anything.

Reply:
@awni I just double checked, you are right. It's using kaiming_uniform to initialize lora a. but, I am not sure if that will cause such differences.
FYI: https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/layer.py#L120

Reply:
It's not inconceivable the initialization would make a difference. I can't tell just from a glance at that code what the scale of the initialization is. But if the vector started out with larger numbers in general that could offset the smaller scale.

Then again there could be a totally different explanation..